### DISCUSSION TITLE
Question:  IL Metadata for optional parameter of type `DateTime`

### DISCUSSION THREAD

**Discussion by HaloFour** [dotnet/csharplang#1236] [General]
https://github.com/dotnet/corefx/issues/26164

The following currently causes a `System.FormatException` due to a bug in the CLR:

```cs
public class SomeType
{
    public void Method(DateTime timestamp = default(DateTime)) { }
}

class Program
{
    static void Main()
    {
        var parameter = typeof(SomeType).GetMethod("Method").GetParameters()[0];
        var defaultValue = parameter.DefaultValue; // exception thrown here
    }
}
```

This is caused by a bug that has been located [here](https://github.com/dotnet/coreclr/blob/d2f2e81294a80a1b463ea185decd07e49e67f976/src/mscorlib/src/System/Reflection/MdConstant.cs#L97-L98) where the CLR expects to see optional `DateTime` parameters with optional values annotated with the `System.Runtime.CompilerServices.DateTimeConstantAttribute` attribute.

This is obviously a bug in the CLR and should be fixed.  The question would be should the C# compiler also emit this attribute to "smooth over" the bug on existing frameworks?  Would that require a change to the language specification or would that be considered an implementation detail?

**Comment by jnm2** [comment]
Yeah, I ran into this too.

**Comment by stakx** [comment]
FYI, I've submitted a bugfix for the CoreCLR (https://github.com/dotnet/coreclr/pull/17877). If this gets merged, I'll try to see whether the fix can be backported to the CLR as well.
